#!/usr/bin/env Rscript

### Author: Juan Hurtado

### This program adds protein type to the table of
### gene/transcript/protein-ids correspondences.
###
### The type of protein (complete, 5prime_partial,
### 3prime_partial, or internal) is added for each predicted
### protein to the table of gene/transcript/protein-ids
### correspondences.
### Each protein type is obtained from sequences attributes
### of the fasta file generated by transdecoder.
### These sequences should be first filtered and revised.
###
### Run this script from Terminal with, e.g.,:
###
### $ Rscript --vanilla annotate_correspondences.R \
###     -a filtered_transdecoder.pep.fasta \
###     -b selected_gn_tr_pp.txt \
###     -o gn_tr_pp.annotated.txt
###


### LIBRARIES
library("optparse")
library("stringr")
library("seqinr")


### ARGUMENTS
option_list <- list(
  make_option(c("--proteins", "-a"), default="filtered_transdecoder.pep.fasta",
              help = "name of the file with the (filtered and revised) transdecoder predicted peptides"),
  make_option(c("--correspondences", "-b"), default="selected_gn_tr_pp.txt",
              help = "gn/tr/pp/ppLength correspondence table file"),
  make_option(c("--output", "-o"), default="gn_tr_pp.annotated.txt",
              help = "name of the annotated correspondence file")
)
opt <- parse_args(OptionParser(option_list=option_list))

# Validating arguments
if (is.null(opt$proteins) || !file.exists(opt$proteins)) {
	stop(paste0("Proteins input file not provided or does not exist. --proteins ", opt$proteins))
}
if (is.null(opt$correspondences) || !file.exists(opt$correspondences)) {
	stop(paste0("Correspondences input file not provided or does not exist. --correspondences ", opt$correspondences))
}
if (!(length(opt$output) && nzchar(opt$output))) {
	stop(paste0("Output can not be empty. Please, choose a file name or skip the option."))
}

### CONSTANTS
input1_file <- opt$proteins
input2_file <- opt$correspondences
output_file <- opt$output

### MAIN PROGRAM
proteins        <- read.fasta(file=input1_file, as.string=TRUE)
correspondences <- read.table(input2_file, header=TRUE)
N <- length(correspondences[,1])
correspondences$ppType <- character(N)
for (i in 1:N) {
	pos <- which(names(proteins) == correspondences$ppID[i])
	x <- attr(proteins[[pos]], "Annot")
	correspondences$ppType[i] <- sub(".*?type:(.*?) len.*", "\\1", x)
	print(paste0("Loop progress ", round(i*100/N, 4), "%"))
}
write.table(correspondences, file=output_file, row.names=FALSE, col.names=TRUE, sep="\t", quote=FALSE)
